modules/diskq/logqueue-disk.c

```diff
@@ -203,4 +225,61 @@ log_queue_disk_init_instance(LogQueueDisk *self, DiskQueueOptions *options, cons
   self->super.type = log_queue_disk_type;
 
   self->qdisk = qdisk_new(options, qdisk_file_id);
+  _set_serializer(self);
+}
+
+static gboolean
+_request_serializer(LogQueueDisk *self, SerializeArchive *sa, GError **error)
+{
+  gpointer args[] = {sa, error};
+  return qdisk_start_serialize(self->qdisk, args);
+}
+
+static gboolean
+_serialize_msg(LogQueueDisk *self, LogMessage *msg, SerializeArchive **sa)
+{
+  if (self->serialize_msg)
+    return self->serialize_msg(msg, *sa);
+
+  return FALSE;
+}
+
+static gboolean
+_overwrite_with_real_record_length(LogQueueDisk *self, GString *serialized, GError **error)
+{
+  gpointer args[] = {serialized, error};
+  return qdisk_end_serialize(self->qdisk, args);
+}
+
+gboolean
+log_queue_disk_serialize_msg(LogQueueDisk *self, LogMessage *msg, GString *serialized)
+{
+  GError *error = NULL;
+  SerializeArchive *sa = serialize_string_archive_new(serialized);
+
+  /* Leave space for the real record_length for later */
+  if (!_request_serializer(self, sa, &error))
```

Comment:
What does "_request" mean in this context? Can you suggest a better name? By reading it I don't really have a clue what I should expect there.

Reply:
looking at the actual code, I can see that this is what emits a record length, e.g. it implements the framing used within the on-disk format. With that in mind this should be something like _prepare_frame. Also, I don't really see whether we need to delegate this back to the qdisk layer. If we do, it could simple prepend the length prior to calling this function that way we could straighten the execution flow. For me this is a bit ziggy-zaggy now.

I would check two alternatives:
1) define which layer has responsibility over what: qdisk is about the disk file format, LogQueueDisk is about the logical representation + the LogQueue glue. Where do we want to delegate the fact that the content in the file is framed. qdisk seems a logical choice, but I don't know whether LogQueueDisk has already been infested by this knowledge or not.

2) if the answer is that we want to delegate framing to qdisk, I would probably implement the framing even before calling into this function. The sole responsibility of this function is to produce a string representation of a LogMessage and qdisk should take it from there. _start_serialize and _end_serialize should not be managed from right within this function.

That's my $.02

Reply:
I believe the ziggy-zaggy way of executing the serialization has been fixed.

Reply:
I'm currently writing proper unit tests for `QDisk`, and it can be seen clearly there that QDisk should be responsible for its framing format. Unfortunately,  its `push_tail()` method allows unframed data, but `pop_head()` expects the framed format, which is a problem.

I will clarify these problems in a separate PR.